pkg/controller/daemon/daemoncontroller.go

```diff
@@ -763,6 +784,12 @@ func (dsc *DaemonSetsController) syncDaemonSet(key string) error {
 	if err != nil {
 		return fmt.Errorf("unable to retrieve ds %v from store: %v", key, err)
 	}
+	// DeepCopy so we don't mutate objects in the cache.
+	cp, err := api.Scheme.DeepCopy(ds)
```

Comment:
We can avoid deep-copying unconditionally. Simply get the live object before L807 where we do the initial deletiontimestamp!=nil check and check the live object instead of the cached object.